fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -1013,18 +1028,49 @@ protected OptExpression queryBasedRewrite(RewriteContext rewriteContext, ScalarO
                 return null;
             }
 
-            // add filter to op
-            Operator.Builder builder = OperatorBuilderFactory.build(queryExpression.getOp());
-            builder.withOperator(queryExpression.getOp());
-            builder.setPredicate(queryCompensationPredicate);
-            Operator newQueryOp = builder.build();
-            OptExpression newQueryExpr = OptExpression.create(newQueryOp, queryExpression.getInputs());
+            OptExpression newQueryExpr = pushdownPredicatesForJoin(queryExpression, queryCompensationPredicate);
             deriveLogicalProperty(newQueryExpr);
             return newQueryExpr;
+
         }
         return null;
     }
 
+    // pushdown predicates on join nodes
+    // the OptExpression will be modified in place
+    private OptExpression pushdownPredicatesForJoin(OptExpression optExpression, ScalarOperator predicate) {
+        if (!(optExpression.getOp() instanceof LogicalJoinOperator)) {
+            if (predicate != null) {
+                // predicate can not be pushdown, we should add it it optExpression
+                Operator.Builder builder = OperatorBuilderFactory.build(optExpression.getOp());
+                builder.withOperator(optExpression.getOp());
+                builder.setPredicate(predicate);
+                Operator newQueryOp = builder.build();
+                return OptExpression.create(newQueryOp, optExpression.getInputs());
+            } else {
+                return optExpression;
+            }
+        }
+        OptExpression newJoin = doPushdownPredicate(optExpression, predicate);
+        // pushdown predicates in children
+        List<OptExpression> children = Lists.newArrayList();
```

Comment:
If there are other operator between join and scan, could this ensure psuh predicate to scan?

Reply:
In the mv rewrite case, there are only join and scan operator.